### DISCUSSION TITLE
HCT ( from Google Material 3)

### DISCUSSION THREAD

**Discussion by qinyiyao** [Evercoder/culori#170] [Ideas]
https://material.io/blog/science-of-color-design
https://github.com/material-foundation/material-color-utilities
![Gradients in HCT respect colorfulness and maintain legibility. From material.io](https://lh3.googleusercontent.com/459crB0shmfQBhnWHtw5UPjX1x42hQ_1GNBxOjiwYootCbqzxno4rleDZ4e8ceuB1RDqyvyrtJTP7QkANIreOmp6Jwefk_yNGL42ZDSnXXCHmuXpvA=w1064-v0)

**Comment by NateBaldwinDesign** [comment]
HCT is just CAM16 with lightness from LAB (which should already be very comparable). I would almost rather see CAM16 implemented over HCT as it is not clearly explained the benefits or purpose of its creation over the existing scientific color appearance model.

**Reply by Artoria2e5** [reply]
See also https://github.com/material-foundation/material-color-utilities/issues/112 -- my question-issue about why they decided to use the Lab lightness instead of what CAM16 already has. In any case, you can't do HCT without CAM16.

**Reply by danburzo** [reply]
I think a bit of the rationale is [described here](https://github.com/material-foundation/material-color-utilities/blob/main/typescript/hct/hct.ts#L18-L32): WCAG defines the contrast ratio based on the Y in XYZ, and L* is a perceptually linear mapping of Y that has the interesting property that a constant ΔL* produces a relatively constant contrast ratio between the colors, regardless of L*. So it does indeed look like ΔL* = 50 reliably produces a contrast ratio greater than 4.5 to 1: 

<img width="1161" alt="A Desmos plot of the WCAG contrast ratio function based on a constant difference in CIELAB lightness shows an almost straight line" src="https://github.com/Evercoder/culori/assets/205375/f9cf849d-a8f7-48fb-aceb-798edeb5301e">

(For brevity, I’ve omitted the linear part of CIELAB transfer function and used `x => x ** 3`)

I’m curious, however, whether the J in CAM16-UCS has similar properties. If it doesn’t, that would explain their choice.

**Comment by danburzo** [comment]
Thanks for submitting the idea, @qinyiyao. I've kept some bookmarks for HCT but didn't have the chance to catch up with them. I agree with @NateBaldwinDesign — if anything, if CAM16 is implemented, support for HCT is a couple of functions away.

I haven't done a close reading on CAM16 and its uniform color space (CAM16-UCS), but skimming [the paper](https://onlinelibrary.wiley.com/doi/10.1002/col.22131) I was reluctant about tackling it, since culori's API doesn't really have a concept of color space parameters — CIELAB under the D50 illuminant (`lab`) is implemented separately from CIELAB under the D65 illuminant (`lab65`), etc. So probably CAM16-UCS would similarly need to be pinned down to certain viewing conditions if it were to to be used in all the ways the [color spaces](https://culorijs.org/color-spaces/) are used throughout the API.

**Reply by Artoria2e5** [reply]
I *think* the UCS already pins down viewing conditions. https://github.com/colour-science/colour/blob/develop/colour/models/cam16_ucs.py, function `XYZ_to_CAM16UCS` takes nothing but XYZ.

**Reply by danburzo** [reply]
The `XYZ_to_CAM16UCS` function is a partial application of `XYZ_to_UCS_Li2017` that fixes the coefficients for deriving the various related color spaces — UCS (uniform color space) vs. SCD (small color differences) vs. LCD (large color differences). The viewing conditions can be configured when invoking the function or left to their defaults:

https://github.com/colour-science/colour/blob/8c090f93384b4a397f96aa47df1db5dca1a3e99d/colour/appearance/ciecam02.py#L163C1-L171C2

**Comment by Artoria2e5** [comment]
The picture seems very suspect. It's almost like HCT was interpolated in a hue-curve-following fashion, while the Oklab and CAM16 are interpolated in Euclidean space, ignoring the cylindrical transforms everyone knows about. That's apples to oranges, blatant cheating.

```js
// The pic does this
interpolate(['blue', 'red'], 'hct')  // not implemented, this feature request
interpolate(['blue', 'red'], 'oklab')
interpolate(['blue', 'red'], 'cam16-ucs-jab')  // not implemented
// instead of this
interpolate(['blue', 'red'], 'oklch')
interpolate(['blue', 'red'], 'cam16-ucs-jch')  // not implemented
```

Oh no, it comes straight from the original blog post! https://material.io/blog/science-of-color-design

**Reply by danburzo** [reply]
Yes, I think a really good question to answer is whether HCT performs significantly better than Oklch.

**Reply by qinyiyao** [reply]
Very sus indeed. I just found this additional comparison image: https://krita-artists.org/t/a-new-color-space/40242/4
![image](https://github.com/Evercoder/culori/assets/6311717/a2d6bf91-8639-4007-8a3f-7ae5f01ea7e5)
Lossy compression aside, it doesn't look that much different.

**Comment by danburzo** [comment]
Link to the related CAM16 discussion, which is a prerequisite for HCT support. 

- https://github.com/Evercoder/culori/discussions/133 

It remains to be seen if explicit HCT support belongs to the library or is better suited as a (small) userland module.
